lib/logical_solver/presolve.cpp

```diff
@@ -0,0 +1,889 @@
+    int row_index, int col_index, 
+    int variable_coefficient, double constraint_RHS
+  ) {
+    // Check if RHS/coeff is not an integer
+
+    if (static_cast<int>(constraint_RHS) % variable_coefficient != 0) {
+      // If inequality, check if rounding up satisfies upper
+      // bound and implied bounds.
+      if (row_index < inequalities_count_) {
+        int feasible_value = ceil(constraint_RHS/variable_coefficient);
+        if (feasible_value <= upper_bounds_.at(row_index)) {
+          if (checkVariableImpliedBounds(col_index, feasible_value)) {
+            return feasible_value;
+          }
+        }
+      }
+      return kInfinity;
+    }
+    // If RHS/coeff is integer, calculate it and check if 
+    // implied bounds satisfied.
+    int feasible_value = constraint_RHS/variable_coefficient;
+    if (checkVariableImpliedBounds(col_index, feasible_value)) {
+      return feasible_value;
+    }
+    return kInfinity;
+  };
+
+  bool Presolve::checkIsRowFree(int row_index) {
+    if (lower_bounds_.at(row_index) == -kInfinity && upper_bounds_.at(row_index) == kInfinity) {
+      return true;
+    }
+    return false;
+  }
+
+  void Presolve::updateStateFreeRow(int row_index) {
+    presolve_active_rows_.at(row_index) = false;
+    presolve_active_rows_count_ -= 1;
+
+    // Update presolve stack.
+    struct presolve_log log = {row_index, -1, 0};
+    presolve_stack_.push(log);
+  };
+
+  void Presolve::applyFreeRowPostsolve(int row_index, int col_index) {
+    postsolve_active_rows_.at(row_index) = true;
+  };
+
+  void Presolve::updateStateSingletonVariable(
+    int row_index, int col_index
+  ) {
+    presolve_active_rows_.at(row_index) = false;
+    presolve_active_columns_.at(col_index) = false;
+    presolve_active_rows_count_ -= 1;
+    presolve_active_columns_count -= 1;
+
+    // Update presolve stack.
+    struct presolve_log log = {row_index, col_index, 1};
+    presolve_stack_.push(log);
+  };
+
+  void Presolve::applySingletonVariablePostsolve(
+    int row_index, int col_index
+  ) {
+    int variable_coeff = problem_matrix_.at(row_index).at(col_index);
+    double feasibleValueCalculationBound = getFeasibleValueCalculationBound(
+      row_index
+    );
+    double feasible_value;
+
+    if (solve_MIP_) {
+      feasible_value = getVariableFeasibleValueMIP(
+        row_index, col_index, problem_matrix_.at(row_index).at(col_index), 
+        feasibleValueCalculationBound
+      );
+    } else {
+      feasible_value = feasibleValueCalculationBound/variable_coeff;
+    }
+
+    if (feasible_value == kInfinity) {infeasible = true;}
+    else {
+      feasible_solution.at(col_index) = feasible_value;
+      postsolve_active_cols_.at(col_index) = true;
+      // Check if we have found feasible values for all variables in the 
+      // row. If so, check if the row (constraint) is satisfied, and if 
+      // so turn on row in postsolve.
+      if (isRowActivePostsolve(row_index)) {
```

Comment:
Check for guard clauses in your code to avoid nesting if statements i.e. instead of

```
if(a){
  if(b){
    do x;
    return;
  }
  else{
    do y;
    return;
  }
}
do z;
return;
```

try using

```
if(!a){
  do z;
  return;
}
if(b){
  do x;
  return;
}
else{
  do y;
  return;
}
```

Reply:
It's a small thing and might not always be relevant but when you spot it, it can help you avoid pointless operations while still conducting the same number of comparisons even in the worst case.

Reply:
Might do it if I have the time before submitting.